utility/SPIDEV/gpio.cpp

```diff
@@ -27,152 +32,83 @@ GPIO::~GPIO()
-        sleep(1);
-        counter++;
-        if (counter > 10) {
-            throw GPIOException("can't access /sys/class/gpio/gpio%d/direction GPIO pin. check access rights");
-            /*perror("Could not open /sys/class/gpio/gpio%d/direction");
-            exit(0); */
-        }
-    }
-    int l = (DDR == 0) ? fprintf(f, "in\n") : fprintf(f, "out\n");
-    if (!(l == 3 || l == 4)) {
-        fclose(f);
-        throw GPIOException("can't set direction on GPIO pin. check access rights");
-    }
-    /*
-    if (DDR == 0)
-        fprintf(f, "in\n");
-    else
-        printf(f, "out\n");
-    */
-    fclose(f);
+int fd;
 
-    // Caches the GPIO descriptor;
-    sprintf(file, "/sys/class/gpio/gpio%d/value", port);
-    int flags = (DDR == 0) ? O_RDONLY : O_WRONLY;
-    int fd = ::open(file, flags);
-    if (fd < 0) {
-        throw GPIOException("Can't open the GPIO");
+void GPIO::Gopen(int port, int DDR)
+{
+    fd = open(dev_name, O_RDONLY);
+    if (fd >= 0) {
+        close(fd);
     }
     else {
-        cache[port] = fd; // cache the fd;
-        lseek(fd, SEEK_SET, 0);
+        dev_name = "/dev/gpiochip0";
+        fd = open(dev_name, O_RDONLY);
+        if (fd >= 0) {
+            close(fd);
+        }
+        else {
+            throw GPIOException("can't open /dev/gpiochip");
+        }
     }
 }
 
-void GPIO::close(int port)
+void GPIO::Gclose(int port)
 {
-    std::map<int, GPIOfdCache_t>::iterator i;
-    i = cache.find(port);
-    if (i != cache.end()) {
-        close(i->second); // close the cached fd
-        cache.erase(i);   // Delete cache entry
-    }
-    // Do unexport
-    FILE* f;
-    f = fopen("/sys/class/gpio/unexport", "w");
-    if (f != NULL) {
-        fprintf(f, "%d\n", port);
-        fclose(f);
-    }
 }
 
-int GPIO::read(int port)
+int GPIO::Gread(int port)
 {
-    std::map<int, GPIOfdCache_t>::iterator i;
-    int fd;
-    i = cache.find(port);
-    if (i == cache.end()) { // Fallback to open the gpio
-        GPIO::open(port, GPIO::DIRECTION_IN);
-        i = cache.find(port);
-        if (i == cache.end()) {
-            throw GPIOException("can't access to GPIO");
+
+    struct gpiohandle_request rq;
+    struct gpiohandle_data data;
+    int fd, ret;
+    fd = open(dev_name, O_RDONLY);
+    if (fd >= 0) {
+        rq.lineoffsets[0] = port;
+        rq.flags = GPIOHANDLE_REQUEST_INPUT;
+        rq.lines = 1;
+        ret = ioctl(fd, GPIO_GET_LINEHANDLE_IOCTL, &rq);
+        if (ret == -1) {
+            throw GPIOException("Can't get line handle from IOCTL");
+            return ret;
         }
-        else {
-            fd = i->second;
+        close(fd);
+        ret = ioctl(rq.fd, GPIOHANDLE_GET_LINE_VALUES_IOCTL, &data);
+        if (ret == -1) {
+            throw GPIOException("Can't get line value from IOCTL");
+            return ret;
         }
+        close(rq.fd);
+        return ret;
```

Comment:
Shouldn't we be returning `dara.value` here? Returning the success code from `ioctl` makes sense in the case of an error.

Reply:
The driver doesn't use the read() function, but I changed it to use `data.values[0]` as per the example code. Nice catch.